File: src/main/java/wraith/fabricaeexnihilo/modules/infested/InfestedLeavesBlock.java
package wraith.fabricaeexnihilo.modules.infested;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.LeavesBlock;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.random.Random;
import net.minecraft.util.registry.Registry;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.modules.base.Colored;
import wraith.fabricaeexnihilo.util.Color;
import wraith.fabricaeexnihilo.util.Lazy;

public class InfestedLeavesBlock extends LeavesBlock implements Colored, NonInfestableLeavesBlock {
    private final Identifier old;
    private final Lazy<Block> leafBlock;
    
    public InfestedLeavesBlock(Identifier leafBlock, Settings settings) {
        super(settings);
        this.old = leafBlock;
        this.leafBlock = new Lazy<>(() -> Registry.BLOCK.get(leafBlock));
    }
    
    @Override
    public int getColor(int index) {
        return Color.WHITE.toInt();
    }
    
    public Identifier getOld() {
        return old;
    }
    
    public Block getLeafBlock() {
        return leafBlock.get();
    }
    
    @Override
    public boolean hasRandomTicks(BlockState state) {
        return true;
    }
    
    @Override
    public void randomTick(BlockState state, ServerWorld world, BlockPos pos, Random random) {
        super.randomTick(state, world, pos, random);
        if (world.isClient) {
            return;
        }
        InfestedHelper.tryToSpreadFrom(world, pos, FabricaeExNihilo.CONFIG.modules.silkworms.infestedSpreadAttempts);
    }
}

File: src/main/kotlin/exnihilofabrico/modules/infested/InfestedLeavesBlock.kt
package exnihilofabrico.modules.infested

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.modules.base.IHasColor
import exnihilofabrico.util.Color
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.fabricmc.fabric.api.block.FabricBlockSettings
import net.minecraft.block.BlockState
import net.minecraft.block.LeavesBlock
import net.minecraft.server.world.ServerWorld
import net.minecraft.text.Text
import net.minecraft.text.TranslatableText
import net.minecraft.util.math.BlockPos
import java.util.*

class InfestedLeavesBlock(val leafBlock: LeavesBlock, settings: FabricBlockSettings): LeavesBlock(settings.build()), IHasColor {
    override fun getColor(index: Int) = Color.WHITE.toInt()


    override fun hasRandomTicks(blockState_1: BlockState?) = true
    override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        super.randomTick(state, world, pos, random)
        if(world.isClient) return
        InfestedHelper.tryToSpreadFrom(
            world,
            pos,
            ExNihiloFabrico.config.modules.silkworms.infestedSpreadAttempts
        )
    }

    @Environment(EnvType.CLIENT)
    override fun getName(): Text {
        return TranslatableText("block.exnihilofabrico.infested", leafBlock.name)
    }
}